compiler/rustc_parse/src/parser/expr.rs

```diff
@@ -137,6 +137,20 @@ impl<'a> Parser<'a> {
                     self.bump();
                     Ok(self.mk_expr(self.prev_token.span, ExprKind::Err))
                 }
+                None if self.may_recover()
+                    && self.prev_token.is_ident()
+                    && self.token.kind == token::Colon =>
+                {
+                    err.span_suggestion_verbose(
+                        self.prev_token.span.until(self.look_ahead(1, |t| t.span)),
+                        "if this is a parameter, remove the name for the parameter",
+                        "",
+                        Applicability::MaybeIncorrect,
+                    );
+                    self.bump();
+                    err.emit();
+                    Ok(self.mk_expr(self.token.span, ExprKind::Err))
```

Comment:
I think we should try parsing a subsequent expression with a [snapshot](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_parse/parser/struct.Parser.html#method.create_snapshot_for_diagnostic), and then return that expression rather than `ExprKind::Err`.

That should also get rid of the subsequent "missing ," diagnostic below.

Reply:
Will work on this in about 13 hours, thank you for the suggestion.

Reply:
If I understand this correctly, we emit the "if this is a parameter" error. Then try to parse an expression in the snapshot, if an expression is found, it will be returned. Else we return an `ExprKind::Err`?

Reply:
Yes. If the expression parses successfully, then you also need to restore the snapshot: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_parse/parser/struct.Parser.html#method.restore_snapshot

And if the parsing succeeds, you need to cancel the diagnostic that comes from it: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_errors/struct.DiagnosticBuilder.html#method.cancel

Reply:
Nevermind, seems like you figured it out.

Reply:
Thank you for the resources!